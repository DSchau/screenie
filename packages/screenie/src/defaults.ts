import * as path from 'path';
import * as yup from 'yup';

import { ScreenieOptions } from './interfaces';

const schemaBase = yup.object().shape({
  adapter: yup.string().default('adapter-default').required(),
  beforeScreenshot: yup.mixed().default(() => () => null),
  delay: yup.number().default(0).required(),
  height: yup.number().default(450).required(),
  width: yup.number().default(800).required(),
  isMobile: yup.boolean().default(true).required()
});

const schema = schemaBase.concat(yup.object().shape({
  deviceScaleFactor: yup.number(),
  folder: yup.string().required(),
  hasTouch: yup.boolean(),
  isLandscape: yup.boolean(),
  url: yup.string().required()
}));

const getConfigFile = (fileName = '.screenierc.js') => {
  try {
    return require(path.join(process.cwd(), fileName));
  } catch (e) {
    return {};
  }
};

export const defaults = async (...options: Partial<ScreenieOptions>[]): Promise<ScreenieOptions> => {
  const base = await schemaBase.validate() as ScreenieOptions;

  const merged = [].concat(options).concat(getConfigFile()).reduce((merged, opts) => ({
      ...merged,
      ...opts
    }), base);

  return await schema.validate(merged) as ScreenieOptions;
};
