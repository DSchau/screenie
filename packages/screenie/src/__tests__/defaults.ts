import { defaults } from '../defaults';

test('it is a function', () => {
  expect(defaults).toEqual(expect.any(Function));
});

test('it returns defaults', async () => {
  expect(await defaults({ folder: 'screenshots', url: 'https://google.com' })).toMatchSnapshot();
});

test('it allows for customization', async () => {
  expect(await defaults({ delay: 10000, folder: 'screenshots', url: 'https://google.com' })).toMatchSnapshot();
});
