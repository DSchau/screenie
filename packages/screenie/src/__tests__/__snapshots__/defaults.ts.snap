// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it allows for customization 1`] = `
Object {
  "adapter": "adapter-default",
  "beforeScreenshot": [Function],
  "delay": 10000,
  "folder": "screenshots",
  "height": 450,
  "isMobile": true,
  "url": "https://google.com",
  "width": 800,
}
`;

exports[`it returns defaults 1`] = `
Object {
  "adapter": "adapter-default",
  "beforeScreenshot": [Function],
  "delay": 0,
  "folder": "screenshots",
  "height": 450,
  "isMobile": true,
  "url": "https://google.com",
  "width": 800,
}
`;
